harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -118,6 +166,96 @@ def _set_random_seeds(cls: Type["PyTorchTrialController"], seed: int) -> None:
         # torch.backends.cudnn.deterministic = True
         # torch.backends.cudnn.benchmark = False
 
+    def report_training_metrics(self, steps_completed, training_metrics):
+        metrics = self._prepare_metrics(training_metrics)
+        self.metric_writer.on_train_step_end(
+            self.steps_completed,
+            metrics["avg_metrics"],
+            metrics["batch_metrics"],
+        )
+        # Only report on the chief worker
+        if self.is_chief:
+            avg_metrics = metrics.get("avg_metrics", {})
+            batch_metrics = metrics.get("batch_metrics", [])
+            self.core_context.train.report_training_metrics(
+                steps_completed=steps_completed, metrics=avg_metrics, batch_metrics=batch_metrics
+            )
+        self.training_metrics = []
+
+    def report_validation_metrics(self, steps_completed, metrics):
+        if self.is_chief:
+            # Check if op complete
+            if self.searcher_op and self._steps_remaining() < 1:
+                searcher_metric = self._validate_searcher_metric(metrics)
+                self.searcher_op.report_completed(searcher_metric)
+
+            self.core_context.train.report_validation_metrics(steps_completed, metrics)
+
+    def on_epoch_start(self, epoch_idx: int):
+        for callback in self.callbacks.values():
+            with self.prof.record_timing(
+                f"callbacks.{callback.__class__.__name__}.on_training_epoch_start"
+            ):
+                sig = signature(callback.on_training_epoch_start)
+                if sig.parameters:
+                    callback.on_training_epoch_start(epoch_idx)
+                else:
+                    logging.warning(
+                        "on_training_epoch_start() without parameters is deprecated"
+                        " since 0.17.8. Please add epoch_idx parameter."
+                    )
+                    callback.on_training_epoch_start()  # type: ignore[call-arg]
+
+    def on_epoch_end(self, epoch_idx: int):
+        if self.context.is_epoch_end():
```

Comment:
IMO it is confusing when a function has a predicate like this.  If it's called `on_epoch_end()` we should only call it when `self.context.is_epoch_end()` is true.

Otherwise you could write valid but confusing code like this:
```python
for batch in dataloader:
    forward_pass()
    backward_pass()
    on_epoch_end()  # every batch, wuuut?
```

You could name it `maybe_epoch_end()` to indicate it is going to detect epoch ends, but IMO it's way simpler to make the caller of `on_epoch_end()` only call it when it _is_ an epoch end, and let `on_epoch_end()` do exactly what it says it does.

Reply:
this was an extraneous call, since `self.context.is_epoch_end()` is already called in the caller of this method. i removed it.